src/Lucene.Net/Util/ManagedScoreDocArray.cs

```diff
@@ -0,0 +1,537 @@
+    public int Length => _length;
+
+    public ManagedScoreDocArray()
+    {
+    }
+
+    public ManagedScoreDocArray(int totalItems, bool hasFields)
+    {
+        _length = totalItems;
+
+        int remainingToAllocate = totalItems;
+        int currentSize = InitialItems;
+
+        while (_segments.Count < StablePhaseSegmentStartIndex && remainingToAllocate > 0)
+        {
+            AllocateSegment(currentSize, hasFields);
+            remainingToAllocate -= currentSize;
+            currentSize *= 2;
+        }
+
+        if (remainingToAllocate > 0)
+        {
+            int stableSegmentsNeeded = (remainingToAllocate + (MaxItemsPerSegment - 1)) >> MaxItemsLog2;
+
+            for (int i = 0; i < stableSegmentsNeeded; i++)
+            {
+                AllocateSegment(MaxItemsPerSegment, hasFields);
+            }
+        }
+
+        // we must point _currentDocs to the last segment and calculate exactly 
+        // how many items are used in that last segment.
+        if (_segments.Count > 0)
+        {
+            var lastIndex = _segments.Count - 1;
+            var lastSeg = _segments[lastIndex];
+
+            _currentDocs = lastSeg.Docs;
+            _currentScores = lastSeg.Scores;
+            _currentSegmentCapacity = lastSeg.Capacity;
+
+            int itemsInPreviousSegments = 0;
+            for (int i = 0; i < lastIndex; i++)
+            {
+                itemsInPreviousSegments += _segments[i].Capacity;
+            }
+
+            _currentSegmentUsed = totalItems - itemsInPreviousSegments;
+
+            // update the struct in the list so readers see the correct limit
+            lastSeg.Used = _currentSegmentUsed;
+        }
+        else
+        {
+            // totalItems was 0
+            _currentDocs = null;
+            _currentScores = null;
+            _currentSegmentCapacity = 0;
+            _currentSegmentUsed = 0;
+        }
+    }
+
+    private void AllocateSegment(int size, bool hasFields)
+    {
+        var docs = ArrayPool<int>.Shared.Rent(size);
+        var scores = ArrayPool<float>.Shared.Rent(size);
+
+        var segment = new Segment
+        {
+            Docs = docs,
+            Scores = scores,
+            Capacity = size,
+            Used = size // default to full, we fix the last one in the constructor
+        };
+
+        if (hasFields)
+            segment.Fields = ArrayPool<IComparable[]>.Shared.Rent(size);
+
+        _segments.Add(segment);
+    }
+
+    public void Add(int doc, float score)
+    {
+        if (_currentSegmentUsed == _currentSegmentCapacity)
+        {
+            EnsureCapacity();
+        }
+
+        // OPTIMIZATION: Unsafe ref arithmetic
+        ref int docsRef = ref MemoryMarshal.GetArrayDataReference(_currentDocs);
+        ref float scoresRef = ref MemoryMarshal.GetArrayDataReference(_currentScores);
+
+        Unsafe.Add(ref docsRef, _currentSegmentUsed) = doc;
+        Unsafe.Add(ref scoresRef, _currentSegmentUsed) = score;
```

Comment:
Does it make sense to do this for just one assignment per array? I remember a case from Sparrow where `MemoryMarshal.GetReference` was very costly.